### ISSUE TITLE
[Feature Request] Option to enable MC6839 Floating Point ROM package within address space

### ISSUE THREAD

**Issue by tomcircuit** [jimbro1000/online6809#25]
I'm experimenting with the Motorola (IEEE-754-ish) Floating Point package, and it would be really convenient to use online6809 to do so. A means to enable the 8K FP ROM image to overlay some (fixed) part of the emulated address space would be wonderful. In this way, the FP functions could be called from ASM source programs.

The floating point ROM image (binary) can be found [here](https://github.com/jefftranter/6809/tree/master/sbc/mc6839/bin) for example.

**Comment by jimbro1000** [comment]
Rather than simply providing a copy of the MC6839 ROM I would propose the ability to load a rom image into memory from file (or even just pasting the hex/bin into a form). The emulator has no concept of rom vs ram or a chip select mechanism so there is no pre-defined address space to achieve compatibility with an existing hardware design. I'm assuming that the rom only uses relative addressing and re-entrant coding so it shouldn't matter where the image exists in memory

**Comment by tomcircuit** [comment]
Yes on all counts! Loading an image from a file would be ideal, and S19 is the norm for Motorola, of course. The FP ROM is 8K, and yes it's able to be located anywhere within the 6809's address space. Obviously, overlaying the vectors wouldn't be ideal.

**Comment by jimbro1000** [comment]
So I'm going to propose a binary loader where you provide the base address and a block of hex data. The loader will be "dumb" in that it provides no checking of fields beyond being a valid base address ($0000-$FFFF) and valid hexadecimal data.

If the address exceeds $FFFF during loading it will just wrap around to $0000

If you overwrite the hardware vectors it is your responsibility to make sure it is viable (the code already does some slightly mysterious stuff around this...

I'm not sure yet what the disassembler behaviour should or would be (its been a while)

The alternative is to include a machine monitor that sucks up the contents of a form field when executed, this would be closer to real world but not helpful for ROM like images.

I have an alternative emulator that I wrote before discovering online6809 that would more closely suit the requirement as it allows you to define a memory map but it doesn't include all the extra features (and I seem to remember it being slightly broken). I may see if I can shoehorn some of that design into this

**Comment by tomcircuit** [comment]
That would absolutely work - just one question: "hex" data as in ASCII hex?  e.g. DE AD BE EF 10 20 30 40  ? 

S19 would be more helpful, as most 6809 assemblers want to output that format.  There are tools to get to ASCII hex dumps, so that's not awful.  S19 is really easy to parse - let me know if you need help with that!

**Comment by jimbro1000** [comment]
Never tried doing anything with S19 - as far as I can tell at a glance and not looking anything up it is a compressed form and just needs a little processing but ultimately the emulator works on raw hex, not S19. Not sure I agree on the assemblers pushing out S19 though - whenever I've done such work it is binary, labels and annotated source that comes out. Then again I don't do an awful lot of 6809 assembler any more.

Should be relatively simple to detect S19 with a reasonable degree of accuracy so it just works automagically
